fix header length in dcd

kain88-de opened the pull request:
Fixes #1701

Changes made in this Pull Request:
 - the remark information is now consistent. Before we said that the title block is 164 bytes while we wrote 244. 
- DCDReader uses istart information to set the correct time.


PR Checklist
------------
 - [x] make `istart` user selectable and default to `1`
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

jbarnoud commented:
Would there be a way to test that?

kain88-de commented:
I asked @tclick to test out this branch. Once we know it works I can write a test to ensure consistency. We might also be able to rewrite the `readdcd.h` code to accept variable length headers.

tclick commented:
I tested this using the CHARMM script, and it works. However, the thermodynamic calculations differ considerably from the original trajectory. This is due to line 403 in coordinates/DCD.py. If you change `istart=0)` to `istart=1)`, then the previous and the trajectory outputs match.

kain88-de commented:
Do DCD's always start at 1? Should we instead make this an option of the write method to let the user decide?

tclick commented:
In MDA 0.10.0, one was allowed to select the start time. You eventually dropped it, but AFAIK, CHARMM starts at one; I cannot say for NAMD, but VMD writes out a DCD starting at 0. That's why the thermodynamic output matched between your updated version and VMD. By allowing the user to select the start time, then s/he can decide where they want to begin the time.

kain88-de commented:
@tclick  I will make it user selectable and set the default to 1. We anyway pretend to write CHARMM files so it's better if we try to follow it.

kain88-de commented:
@tclick can you test again.

I now set the remarks section to be 240 chars long by default (instead of 160 I had before). I also added an option to set `istart` (default = 1) and added a unit test to check if we store the remark information correctly in the DCD files. This option requires the least changes and keeps the long comments strings we decided to support when we the switched to libdcd. I would like a report if this works though. If there are any problems I switch back to 160 chars.

tylerjereddy commented:
@kain88-de I see the review request -- will probably be Sunday at the earliest, though I'm guessing you're already fresher on the details than I. I do remember this being pretty fragile when we were porting to Python 3.

tclick commented:
I am happy to report that the fix works. Thanks for the change.

kain88-de commented:
@tylerjereddy I'm done now. I appreciate if you have a look on the weekend.

kain88-de commented:
@tclick this PR should have another nice change for you now. The DCDReader now checks the `istart` variable in the DCD header to set the correct time. Before we always started from 0. A side effect of this is that I changed the default of `istart` back to 0 so that the DCDWriter starts writing from time 0, which I think is consistent with the rest of MDAnalysis.

tclick commented:
@kain88-de <https://github.com/kain88-de>, that’s fine. Knowing that I should set `istart=1` for CHARMM is perfectly okay. I understand the need for consistency. Besides, NAMD may honestly start at 0 (although I’m not as familiar with NAMD). You might want to put a note in the docs about this difference, so users can set `istart=1` when wanting to use CHARMM for further work.

I appreciate the work on this; in my current project, I moved away from using CHARMM for certain features because of this bug, and I also realized that the results could be accomplished purely in Python with MDA and numpy (albeit possibly a little slower but with a reduction in external dependencies).

Cordially,
柯明 Timothy H. Click, Ph.D.
Department of Biological Science and Technology
Institute of Bioinformatics and Systems Biology
National Chiao Tung University
208 Lab Building 1, 75 Bo-Ai St.
Dong District, Hsinchu, Taiwan 30062
(R.O.C.)
+886-3-5712121 x56997
tclick@nctu.edu.tw

richardjgowers commented on testsuite/MDAnalysisTests/formats/test_libdcd.py:
```diff
@@ -242,6 +243,22 @@ def test_write_header(tmpdir):
+    with open(testfile, 'rb') as fh:
+        header_bytes = fh.read()
+    # check for magic number
+    assert struct.unpack('i', header_bytes[:4])[0] == 84
```
It's been a while since I had to reverse engineer Fortran binary, but this looks right-
 the magic numbers are the length in bytes of the arrays

kain88-de replied:
Yes, the magic number is the length of the data block. It looks like all data blocks in the DCD file are marked with numbers of the corresponding length at the beginning and end. As I can see the header of a DCD file has 3-4 blocks. 

1. *84 bytes* time information like dt, nsavc, ...
1. *x * 80+4 bytes* remarks
1. *4 bytes* natoms
1. *xx bytes* fixed atoms (may not be written)

tylerjereddy commented on package/CHANGELOG:
```diff
@@ -26,6 +26,8 @@ Fixes
     (Issue #1759)
   * AtomGroup.dimensions now strictly returns a copy (Issue #1582)
   * lib.distances.transform_StoR now checks input type (Issue #1699)
+  * libdcd now write correct length of remark section (Issue #1701)
```
`write` -> `writes`

tylerjereddy commented on package/CHANGELOG:
```diff
@@ -26,6 +26,8 @@ Fixes
   * AtomGroup.dimensions now strictly returns a copy (Issue #1582)
   * lib.distances.transform_StoR now checks input type (Issue #1699)
+  * libdcd now write correct length of remark section (Issue #1701)
+  * DCDReader now reports to correct time based on istart information (PR #1767)
```
`to` -> `the`

tylerjereddy commented on testsuite/MDAnalysisTests/analysis/test_rms.py:
```diff
@@ -165,21 +165,21 @@ def outfile(self, tmpdir):
-        return [[0, 0, 0, 0, 0],
-                [49,   49,   4.6997, 1.9154, 2.7139]]
+        return [[0, 1000, 0, 0, 0],
+                [49,   1049,   4.6997, 1.9154, 2.7139]]
```
Can you clarify why results in the `test_rms` suite are changing because of subtle changes in DCD handling machinery?

tylerjereddy replied:
i.e., if `DCD` test file is being used to seed the tests, that might make sense

kain88-de replied:
This is the changed time. The DCD file doesn't start at time point 0 but rather at time 1000. We have been treating this wrong for some time.

tylerjereddy commented on testsuite/MDAnalysisTests/coordinates/test_memory.py:
```diff
@@ -75,6 +75,8 @@ def reader(self, trajectory):
     def iter_ts(self, i):
         ts = self.universe.trajectory[i]
+        # correct time because memory reader doesn't read the correct time
+        ts.time = ts.frame * self.dt
```
Can you briefly clarify why DCD change is impacting memory reader here?

kain88-de replied:
the memory reader tests are based on the DCD file and assumed it started from 0. This code restores that assumption. The memory reader can't actually deal with a trajectory that doesn't start at 0 at the moment. See the new issue #1769

kain88-de commented on package/MDAnalysis/lib/formats/include/readdcd.h:
```diff
@@ -735,13 +735,16 @@ int write_dcdheader(fio_fd fd, const char *remarks, int N,
   strncpy(title_string, remarks, 240);
+  // Enforce null-termination for long remark strings.
+  // Not a problem for MDAnalysis but maybe for other readers.
+  title_string[239] = '\0';
```
I noticed we have a problem with 0 terminating a string if they are longer then 240 characters. This isn't an issue for writing! It might cause issues for reading with external libraries. Our own cython code seems to handle a missing 0 terminator just fine. There is almost no way for us to test this. The cython library converts the char array into proper python string internally. We never get to see the raw char buffer in python. 

I can again to a test of the expected byte. But nothing more.

kain88-de replied:
I adjusted the tests and docs for this change. This way we can be sure that any DCD written by MDAnalysis doesn't trip over another C library reading DCD.
